### ISSUE TITLE
sampling points　lidar2img问题

### ISSUE THREAD

**Issue by zhaoyangwei123** [MCG-NJU/SparseBEV#64]
@afterthat97 
    B, Q, T, G, P, _ = sample_points.shape  # [B, Q, T, G, P, 3]
    N = 6
    
    sample_points = sample_points.reshape(B, Q, T, G * P, 3)

    # get the projection matrix
    lidar2img = lidar2img[:, :, None, None, :, :]  # [B, TN, 1, 1, 4, 4]
    lidar2img = lidar2img.expand(B, T*N, Q, G * P, 4, 4)
    lidar2img = lidar2img.reshape(B, T, N, Q, G*P, 4, 4)

    # expand the points
    ones = torch.ones_like(sample_points[..., :1])
    sample_points = torch.cat([sample_points, ones], dim=-1)  # [B, Q, GP, 4]
    sample_points = sample_points[:, :, None, ..., None]     # [B, Q, T, GP, 4]
    sample_points = sample_points.expand(B, Q, N, T, G * P, 4, 1)
    sample_points = sample_points.transpose(1, 3)   # [B, T, N, Q, GP, 4, 1]

    # project 3d sampling points to N views
    sample_points_cam = torch.matmul(lidar2img, sample_points).squeeze(-1)  # [B, T, N, Q, GP, 4]

    # homo coord -> pixel coord
    homo = sample_points_cam[..., 2:3]
    homo_nonzero = torch.maximum(homo, torch.zeros_like(homo) + eps)  #third dim is z which need to be 0 in x-y coord
    sample_points_cam = sample_points_cam[..., 0:2] / homo_nonzero  # [B, T, N, Q, GP, 2]   
    您好，我尝试使用您对 sampling points的投影方法，我按照此方法尝试将3D框投影到image上面查看效果，但是发现利用lidar2image之后无论在原图还是load过的图上画出的框子都不对，请问对于3D框您代码中的lidar2image需要有所不同吗？

**Comment by afterthat97** [comment]
没有什么不同啊

**Comment by zhaoyangwei123** [comment]
# Define corners in box local frame
    corners = torch.tensor([
        [0.5, 0.5, 0.5], [0.5, 0.5, -0.5],
        [-0.5, 0.5, 0.5], [-0.5, 0.5, -0.5],
        [0.5, -0.5, 0.5], [0.5, -0.5, -0.5],
        [-0.5, -0.5, 0.5], [-0.5, -0.5, -0.5]
    ], device=device).float()  # (8, 3)

    # Extract dimensions
    dims = box3d[:, :, 3:6]  # (B, N, 3)
    dims = dims.unsqueeze(2).expand(-1, -1, 8, -1)  # (B, N, 8, 3)

    # Extract angles and compute rotation matrices
    sin_theta = box3d[:, :, 6].unsqueeze(-1)
    cos_theta = box3d[:, :, 7].unsqueeze(-1)
    zeros = torch.zeros_like(sin_theta)
    ones = torch.ones_like(sin_theta)
    
    R = torch.stack([
        cos_theta, -sin_theta, zeros,
        sin_theta, cos_theta, zeros,
        zeros, zeros, ones
    ], dim=-1).view(B, N, 3, 3)  # (B, N, 3, 3)
    
    # Apply rotation and translation
    corners = corners.unsqueeze(0).unsqueeze(0) * dims  # (B, N, 8, 3)
    corners_rotated = torch.einsum('bnij,bnmj->bnmi', R, corners)  # (B, N, 8, 3)
    translations = box3d[:, :, :3].unsqueeze(2).expand(-1, -1, 8, -1)  # (B, N, 8, 3)
    vertices = corners_rotated + translations
您好，我发现将gt_box_3d的中心点投过去是正确的，但是我按照这个方式去得到gt_box_3d的八个角点然后投影过去发现这些点和中心点有位移，而且差的很多，请问是我这里对角点的计算造成的吗？

**Comment by zhaoyangwei123** [comment]
作者您好，非常感谢您的工作，我在此基础上加入一些loss的时候遇到一些梯度问题，能加您微信聊一下吗？

**Comment by afterthat97** [comment]
我之前回复过你的邮件，里面有我的微信，你翻翻吧
